File: 2011-fall/Operating System/rw-lock/prototype/writer_pri.py
# - * - coding: UTF-8 - * -

import iothread
import threading

reader_count = 0
writer_count = 0
reader_count_mutex = threading.Lock()
new_reader_mutex = threading.Lock()
writer_count_mutex = threading.Lock()
writer_mutex = threading.Lock()

class ReaderThread(iothread.ReaderThread):
    def do(self):
        global reader_count

        new_reader_mutex.acquire()
        reader_count_mutex.acquire()
        reader_count += 1
        if reader_count == 1:
            writer_mutex.acquire()
        reader_count_mutex.release()
        new_reader_mutex.release()
    
        self.doread()
    
        reader_count_mutex.acquire()
        reader_count -= 1
        if reader_count == 0:
            writer_mutex.release()
        reader_count_mutex.release()

class WriterThread(iothread.WriterThread):
    def do(self):
        global writer_count
        
        writer_count_mutex.acquire()
        writer_count += 1
        if writer_count == 1:
            new_reader_mutex.acquire()
        writer_count_mutex.release()

        writer_mutex.acquire()
        self.dowrite()
        writer_mutex.release()

        writer_count_mutex.acquire()
        writer_count -= 1
        if writer_count == 0:
            new_reader_mutex.release()
        writer_count_mutex.release()

File: 2011-fall/Operating System/rw-lock/program/writer_pri.c
#include <pthread.h>
#include "op.h"
#include "rw.h"

static int reader_count = 0;
static int writer_count = 0;
static pthread_mutex_t reader_count_mutex = PTHREAD_MUTEX_INITIALIZER;
static pthread_mutex_t new_reader_mutex = PTHREAD_MUTEX_INITIALIZER;
static pthread_mutex_t writer_count_mutex = PTHREAD_MUTEX_INITIALIZER;
static pthread_mutex_t writer_mutex = PTHREAD_MUTEX_INITIALIZER;

void writer_pri_reader(int num, int duration) {
    pthread_mutex_lock(&new_reader_mutex);
    pthread_mutex_lock(&reader_count_mutex);
    ++reader_count;
    if (reader_count == 1)
        pthread_mutex_lock(&writer_mutex);
    pthread_mutex_unlock(&reader_count_mutex);
    pthread_mutex_unlock(&new_reader_mutex);

    doread(num, duration);

    pthread_mutex_lock(&reader_count_mutex);
    --reader_count;
    if (reader_count == 0)
        pthread_mutex_unlock(&writer_mutex);
    pthread_mutex_unlock(&reader_count_mutex);
}

void writer_pri_writer(int num, int duration) {
    pthread_mutex_lock(&writer_count_mutex);
    ++writer_count;
    if (writer_count == 1)
        pthread_mutex_lock(&new_reader_mutex);
    pthread_mutex_unlock(&writer_count_mutex);

    pthread_mutex_lock(&writer_mutex);
    dowrite(num, duration);
    pthread_mutex_unlock(&writer_mutex);

    pthread_mutex_lock(&writer_count_mutex);
    --writer_count;
    if (writer_count == 0)
        pthread_mutex_unlock(&new_reader_mutex);
    pthread_mutex_unlock(&writer_count_mutex);
}
